crates/red_knot/src/program/check.rs

```diff
@@ -0,0 +1,284 @@
+use crate::cancellation::CancellationToken;
+use crate::db::SourceDb;
+use crate::files::FileId;
+use crate::lint::Diagnostics;
+use crate::program::Program;
+use rayon::max_num_threads;
+use rustc_hash::FxHashSet;
+use std::num::NonZeroUsize;
+
+impl Program {
+    /// Checks all open files in the workspace and its dependencies.
+    #[tracing::instrument(level = "debug", skip_all)]
+    pub fn check(
+        &self,
+        scheduler: &dyn CheckScheduler,
+        cancellation_token: CancellationToken,
+    ) -> Result<Vec<String>, CheckError> {
+        let check_loop = CheckFilesLoop::new(scheduler, cancellation_token);
+
+        check_loop.run(self.workspace().open_files.iter().copied())
+    }
+
+    /// Checks a single file and its dependencies.
+    #[tracing::instrument(level = "debug", skip(self, scheduler, cancellation_token))]
+    pub fn check_file(
+        &self,
+        file: FileId,
+        scheduler: &dyn CheckScheduler,
+        cancellation_token: CancellationToken,
+    ) -> Result<Vec<String>, CheckError> {
+        let check_loop = CheckFilesLoop::new(scheduler, cancellation_token);
+
+        check_loop.run([file].into_iter())
+    }
+
+    #[tracing::instrument(level = "debug", skip(self, context))]
+    fn check_file_with_context(
```

Comment:
This name seems confusing -- it doesn't really clarify how it is different from `check_file`. It's more like `actually_check_file`. The fact that it takes a context is clear from the signature, doesn't really add anything in the name. Maybe just `do_check`?

Right now it doesn't even really need the context (`CheckFileTask::run` could check for cancellation), but in future when we do more complex and slower checks, we'll probably want to do more cancellation checks internally here, not just one at the start?

Reply:
My idea is to make cancellation checks on every query, and expose an API so that we can perform checks even inside of queries.